File: leetcode/117-populating-next-right-pointers-in-each-node-ii.py
# Definition for binary tree with next pointer.
# class TreeLinkNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
#         self.next = None

class Solution(object):
    def connect(self, root):
        """
        :type root: TreeLinkNode
        :rtype: nothing
        """
        i, start = root, None
        while i:
            j, k = i, None
            while j:
                if k is None:
                    if j.left is not None:
                        start = k = j.left
                        if j.right is not None:
                            k.next = j.right
                            k = k.next
                    elif j.right is not None:
                        start = k = j.right
                else:
                    if j.left is not None:
                        k.next = j.left
                        k = k.next
                    if j.right is not None:
                        k.next = j.right
                        k = k.next
                j = j.next
            i = start
            start = None

File: leetcode/117-populating-next-right-pointers-in-each-node-ii.cpp
/**
 * Definition for binary tree with next pointer.
 * struct TreeLinkNode {
 *  int val;
 *  TreeLinkNode *left, *right, *next;
 *  TreeLinkNode(int x) : val(x), left(NULL), right(NULL), next(NULL) {}
 * };
 */
class Solution {
 public:
  void connect(TreeLinkNode *root) {
    for (TreeLinkNode *i = root, *start = nullptr; i;
         i = start, start = nullptr) {
      for (TreeLinkNode *j = i, *k = nullptr; j; j = j->next) {
        if (k) {
          if (j->left) k->next = j->left, k = k->next;
          if (j->right) k->next = j->right, k = k->next;
        } else {
          if (j->left) {
            start = k = j->left;
            if (j->right)
              k->next = j->right, k = k->next;
          } else if (j->right)
            start = k = j->right;
        }
      }
    }
  }
};
